compiler/ast/types.nim

```diff
@@ -1398,3 +1400,123 @@ proc isObjLackingTypeField*(typ: PType): bool {.inline.} =
+  # distinct types, so use ``sameBackendType`` instead of ``sameType``
+  if isInd and sameBackendType(t, search):
+    # the searched for type is reachable from itself through a ref indirection
+    return true
+
+  let trace = getAttachedOp(g, t, attachedTrace)
+  if trace != nil and sfOverriden in trace.flags:
+    # the type has a custom trace hook. We take it as implying that reasoning
+    # about the type is not possible, and have to assume that a cycle through
+    # the type is possible
+    return true
+
+  case t.kind
+  of tyRef:
+    let base = t.base.skipTypes(abstractInst - {tyTypeDesc})
+    if base.kind == tyObject and not isFinal(base):
+      # a polymorphic ``ref``. We have to assume that the dynamic type contains
+      # the `startId` type somewhere
+      result = tfAcyclic notin base.flags
+    else:
+      # a static ``ref``. Remember that we're analysing a type reachable
+      # through a ref indirection and check the element type
+      result = check(marker, g, base, search, isInd=true)
+
+  of tyProc:
+    # since the env type is dynamic (only known at run-time), we don't know if a
+    # cycle to the type we're searching for is possible (or not possible) -- we
+    # have to assume that it's possible
+    result = t.callConv == ccClosure
+  else:
+    result = trav(marker, g, t, search, isInd)
+
+proc travNode(marker: var IntSet, g: ModuleGraph, n: PNode, search: PType,
+              isInd: bool): bool =
+  ## Traverses the fields of the given record `n` and checks if the type
+  ## identified by `startId` is reachable through one of them. Returns the
+  ## result.
+  result = false
+  if n != nil:
+    case n.kind
+    of nkSym:
+      # cursor fields are non-owning -- they can't keep references alive and
+      # thus can't cause a reference cycle
+      if sfCursor notin n.sym.flags:
+        result = check(marker, g, n.typ, search, isInd)
+    of nkWithSons:
+      for it in n.items:
+        result = travNode(marker, g, it, search, isInd)
+        if result: return
+    of nkWithoutSons - {nkSym}:
+      discard "not relevant"
+
+proc trav(marker: var IntSet, g: ModuleGraph, t: PType, search: PType,
+          isInd: bool): bool =
+  let t = t.skipTypes(abstractInst-{tyTypeDesc})
+
+  case t.kind
+  of tyTuple, tySequence, tyArray:
+    # value types that can keep something alive
+    for i in 0..<t.len:
+      result = check(marker, g, t[i], search, isInd)
+
+  of tyObject:
+    # special handling for objects
+    if containsOrIncl(marker, t.id): # prevent recursion
+      return false
+
+    if t.base != nil and
+       trav(marker, g, t.base.skipTypes(abstractPtrs), search, isInd):
+      return true
+
+    result = travNode(marker, g, t.n, search, isInd)
+  of tyProc:
+    assert t.callConv != ccClosure
+    result = false
+  of tyOpenArray, tyVarargs:
+    # these are views; they don't own their items and thus can't keep them
+    # alive
+    result = false
+  of tyRef:
+    assert false
+  else:
```

Comment:
I'd recommend making these `case` checks exhaustive, withtout a catch all `else`, in order to avoid errors due to changes in `TTypeKind`.